### PR TITLE
Merge/foundation release/1.12.2

### PR DESCRIPTION
**Pull request by ziogaschr**
Merges [ethereum/go-ethereum branch release/1.12.2](https://github.com/ethereum/go-ethereum/releases/tag/v1.12.2) to core-geth master.

# Notes
---
- Do you agree on replacing Clique with Catalyst for `--dev` mode? We still allow use `--dev.pow` for ethash. 9d4ea2e83fc21b41135984363a06334aff106055
- This change follows the go-ethereum interfaces, where a Block header can have `withdrawalsRoot` while the Block itself can have `withdrawals`.  23da7ca9d6ae460324a766a0830b4275b8dc9764
...

# TODO

- [x] Re-implement `RemovePendingTransaction`, as it's not working now that blobpool has been added Ref: https://github.com/etclabscore/core-geth/pull/203
    - :arrow_right: We decided to remove this method since it's not well supported by all transaction types anymore. (https://github.com/etclabscore/core-geth/pull/560#discussion_r1311523912)
- [x] Test that `--dev` mode works https://github.com/etclabscore/core-geth/pull/560#issuecomment-1698943446

### CODE REVIEW

File: eth/api_debug.go [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+			return uint64(i), nil
+		}
+	}
+	return 0, errors.New("no state found")
+}
+
+// RemovePendingTransaction removes a transaction from the txpool.
+// It returns the transaction removed, if any.
+func (api *DebugAPI) RemovePendingTransaction(hash common.Hash) (*types.Transaction, error) {
+	// TODO (ziogaschr): add support for RemoveTx in bloppool
```
Comment (ziogaschr) [inline eth/api_debug.go]: @meowsbits this is a leftover, I will have a look to it. Have you already checked into blobpool?

Comment (ziogaschr) [inline eth/api_debug.go]:
```suggestion
	// TODO (ziogaschr): add support for RemoveTx in blobpool
```

**Comment by ziogaschr** [comment]
Note: I still want to test the functionality while running in `--dev` mode

File: cmd/clef/README.md [hunk 2]
```diff
@@ -29,7 +29,7 @@ GLOBAL OPTIONS:
    --loglevel value        log level to emit to the screen (default: 4)
    --keystore value        Directory for the keystore (default: "$HOME/.ethereum/keystore")
    --configdir value       Directory for Clef configuration (default: "$HOME/.clef")
-   --chainid value         Chain id to use for signing (1=mainnet, 2=Morden (disused), 3=ropsten, 4=rinkeby, 5=Goerli) (default: 1)
+   --chainid value         Chain id to use for signing (1=mainnet, 5=Goerli) (default: 1)
```
Comment (diega) [inline cmd/clef/README.md]:
```suggestion
   --chainid value         Chain id to use for signing (1=foundation, 5=Goerli, 61=classic, 63=Mordor) (default: 1)
```

File: eth/api_debug.go (hunk 1 above)
Comment (ziogaschr) [reply eth/api_debug.go]: We decided to remove support for it.

File: cmd/devp2p/README.md [hunk 3]
```diff
@@ -44,7 +44,7 @@ set to standard output. The following filters are supported:
 - `-limit <N>` limits the output set to N entries, taking the top N nodes by score
 - `-ip <CIDR>` filters nodes by IP subnet
 - `-min-age <duration>` filters nodes by 'first seen' time
-- `-eth-network <mainnet/rinkeby/goerli/sepolia>` filters nodes by "eth" ENR entry
+- `-eth-network <mainnet/goerli/sepolia>` filters nodes by "eth" ENR entry
```
Comment (diega) [inline cmd/devp2p/README.md]:
```suggestion
- `-eth-network <classic/mordor/mainnet/goerli/sepolia>` filters nodes by "eth" ENR entry
```

**Review by diega — APPROVED** [review]
This looks really well. Just made two microscopical comments

**Comment by meowsbits** [comment]
Two new fields are added to the Genesis type: `ExcessBlobGas` and `BlobGasUsed`. While these, along with `BaseFee`, are relevant to ETHereum chains, they are not relevant for all chains that core-geth supports, resulting in empty (`null`) fields. 

What do you think about adding `omitempty` tags for these fields? 

```diff
diff --git a/params/types/genesisT/gen_genesis.go b/params/types/genesisT/gen_genesis.go
index 0136309ee6..12f3d10b05 100644
--- a/params/types/genesisT/gen_genesis.go
+++ b/params/types/genesisT/gen_genesis.go
@@ -33,9 +33,9 @@ func (g Genesis) MarshalJSON() ([]byte, error) {
 		Number        math.HexOrDecimal64                         `json:"number"`
 		GasUsed       math.HexOrDecimal64                         `json:"gasUsed"`
 		ParentHash    common.Hash                                 `json:"parentHash"`
-		BaseFee       *math.HexOrDecimal256                       `json:"baseFeePerGas"`
-		ExcessBlobGas *math.HexOrDecimal64                        `json:"excessBlobGas"`
-		BlobGasUsed   *math.HexOrDecimal64                        `json:"blobGasUsed"`
+		BaseFee       *math.HexOrDecimal256                       `json:"baseFeePerGas,omitempty"`
+		ExcessBlobGas *math.HexOrDecimal64                        `json:"excessBlobGas,omitempty"`
+		BlobGasUsed   *math.HexOrDecimal64                        `json:"blobGasUsed,omitempty"`
 	}
 	var enc Genesis
 	enc.Config = g.Config
diff --git a/params/types/genesisT/genesis.go b/params/types/genesisT/genesis.go
index f2be54ec74..97676daee0 100644
--- a/params/types/genesisT/genesis.go
+++ b/params/types/genesisT/genesis.go
@@ -55,9 +55,9 @@ type Genesis struct {
 	Number        uint64      `json:"number"`
 	GasUsed       uint64      `json:"gasUsed"`
 	ParentHash    common.Hash `json:"parentHash"`
-	BaseFee       *big.Int    `json:"baseFeePerGas"` // EIP-1559
-	ExcessBlobGas *uint64     `json:"excessBlobGas"` // EIP-4844
-	BlobGasUsed   *uint64     `json:"blobGasUsed"`   // EIP-4844
+	BaseFee       *big.Int    `json:"baseFeePerGas,omitempty"` // EIP-1559
+	ExcessBlobGas *uint64     `json:"excessBlobGas,omitempty"` // EIP-4844
+	BlobGasUsed   *uint64     `json:"blobGasUsed,omitempty"`   // EIP-4844
 }
 
 func (g *Genesis) GetElasticityMultiplier() uint64 {
```

**Comment by ziogaschr** [comment]
> Two new fields are added to the Genesis type: `ExcessBlobGas` and `BlobGasUsed`. While these, along with `BaseFee`, are relevant to ETHereum chains, they are not relevant for all chains that core-geth supports, resulting in empty (`null`) fields.
> 
> What do you think about adding `omitempty` tags for these fields?
> 
> ```diff
> diff --git a/params/types/genesisT/gen_genesis.go b/params/types/genesisT/gen_genesis.go
> index 0136309ee6..12f3d10b05 100644
> --- a/params/types/genesisT/gen_genesis.go
> +++ b/params/types/genesisT/gen_genesis.go
> @@ -33,9 +33,9 @@ func (g Genesis) MarshalJSON() ([]byte, error) {
>  		Number        math.HexOrDecimal64                         `json:"number"`
>  		GasUsed       math.HexOrDecimal64                         `json:"gasUsed"`
>  		ParentHash    common.Hash                                 `json:"parentHash"`
> -		BaseFee       *math.HexOrDecimal256                       `json:"baseFeePerGas"`
> -		ExcessBlobGas *math.HexOrDecimal64                        `json:"excessBlobGas"`
> -		BlobGasUsed   *math.HexOrDecimal64                        `json:"blobGasUsed"`
> +		BaseFee       *math.HexOrDecimal256                       `json:"baseFeePerGas,omitempty"`
> +		ExcessBlobGas *math.HexOrDecimal64                        `json:"excessBlobGas,omitempty"`
> +		BlobGasUsed   *math.HexOrDecimal64                        `json:"blobGasUsed,omitempty"`
>  	}
>  	var enc Genesis
>  	enc.Config = g.Config
> diff --git a/params/types/genesisT/genesis.go b/params/types/genesisT/genesis.go
> index f2be54ec74..97676daee0 100644
> --- a/params/types/genesisT/genesis.go
> +++ b/params/types/genesisT/genesis.go
> @@ -55,9 +55,9 @@ type Genesis struct {
>  	Number        uint64      `json:"number"`
>  	GasUsed       uint64      `json:"gasUsed"`
>  	ParentHash    common.Hash `json:"parentHash"`
> -	BaseFee       *big.Int    `json:"baseFeePerGas"` // EIP-1559
> -	ExcessBlobGas *uint64     `json:"excessBlobGas"` // EIP-4844
> -	BlobGasUsed   *uint64     `json:"blobGasUsed"`   // EIP-4844
> +	BaseFee       *big.Int    `json:"baseFeePerGas,omitempty"` // EIP-1559
> +	ExcessBlobGas *uint64     `json:"excessBlobGas,omitempty"` // EIP-4844
> +	BlobGasUsed   *uint64     `json:"blobGasUsed,omitempty"`   // EIP-4844
>  }
>  
>  func (g *Genesis) GetElasticityMultiplier() uint64 {
> ```

Makes sense to me too. Might be that a test can break, in case they check for this fields, but maybe it makes sense to modify those tests as well.

**Comment by meowsbits** [comment]
Yea, my rationale for removing them is mostly based on the pattern of extensive use of `omitempty` in the chain config types, which is obviously a closely related data type. So I would infer and assume that similar tests would be used for both data types, and that therefore checks for the existence or absence of fields should have similar expectations (permitting, or even expecting, that zero-value fields are omitted).

**Review by meowsbits — APPROVED** [review]

**Pull request merged**